### PR TITLE
feat(pubsub): implement per-call options for `Subscriber`

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #7689

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10043)

### CODE REVIEW

File: google/cloud/pubsub/subscriber_test.cc [hunk 1]
```diff
@@ -76,6 +92,103 @@ TEST(SubscriberTest, SubscribeWithOptions) {
...
+    EXPECT_EQ(current.get<TestOptionA>(), "override-a");
+    EXPECT_EQ(current.get<TestOptionB>(), "override-b2");
+    EXPECT_EQ(current.get<TestOptionC>(), "test-c");
+    return make_ready_future(Status{});
+  });
+
+  Subscriber subscriber(mock, Options{}.set<TestOptionA>("override-a"));
+  ASSERT_STATUS_OK(subscriber
+                       .Subscribe([](Message const&, AckHandler const&) {},
+                                  Options{}.set<TestOptionB>("override-b1"))
```
Comment (devbww) [inline google/cloud/pubsub/subscriber_test.cc:183]: Perhaps these should also be setting `TestOptionA` too, to check that call overrides client.

File: google/cloud/pubsub/subscriber.cc [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace google {
+namespace cloud {
+namespace pubsub {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+Subscriber::Subscriber(std::shared_ptr<SubscriberConnection> connection,
+                       Options opts)
+    : connection_(std::move(connection)),
+      options_(google::cloud::internal::MergeOptions(std::move(opts),
```
Comment (devbww) [inline google/cloud/pubsub/subscriber.cc:25]: Q.  Do we like `google::cloud::` qualifications when we're already in `google::cloud`?

File: google/cloud/pubsub/subscriber_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/subscriber_test.cc]: Done

File: google/cloud/pubsub/subscriber.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/subscriber.cc]: Fixed.

**Review by devbww — APPROVED** [review]

**Pull request merged**